extensions/src/AWSSDK.Extensions.NETCore.Setup/ClientFactory.cs

```diff
@@ -203,67 +206,191 @@ private static AWSCredentials CreateCredentials(ILogger logger, AWSOptions optio
+            }
+            if (defaultConfig.LogMetrics.HasValue)
+            {
+                config.LogMetrics = defaultConfig.LogMetrics.Value;
+            }
+            if (defaultConfig.LogResponse.HasValue)
+            {
+                config.LogResponse = defaultConfig.LogResponse.Value;
+            }
+            if (defaultConfig.MaxErrorRetry.HasValue)
+            {
+                config.MaxErrorRetry = defaultConfig.MaxErrorRetry.Value;
+            }
+            if (defaultConfig.ProgressUpdateInterval.HasValue)
+            {
+                config.ProgressUpdateInterval = defaultConfig.ProgressUpdateInterval.Value;
+            }
+            if (defaultConfig.RequestMinCompressionSizeBytes.HasValue)
+            {
+                config.RequestMinCompressionSizeBytes = defaultConfig.RequestMinCompressionSizeBytes.Value;
+            }
+            if (defaultConfig.ResignRetries.HasValue)
+            {
+                config.ResignRetries = defaultConfig.ResignRetries.Value;
+            }
+            if (defaultConfig.RetryMode.HasValue)
+            {
+                config.RetryMode = defaultConfig.RetryMode.Value;
+            }
+            if (defaultConfig.ThrottleRetries.HasValue)
+            {
+                config.ThrottleRetries = defaultConfig.ThrottleRetries.Value;
+            }
+            if (defaultConfig.Timeout.HasValue)
+            {
+                config.Timeout = defaultConfig.Timeout.Value;
+            }
+            if (defaultConfig.UseAlternateUserAgentHeader.HasValue)
+            {
+                config.UseAlternateUserAgentHeader = defaultConfig.UseAlternateUserAgentHeader.Value;
+            }
+            if (defaultConfig.UseDualstackEndpoint.HasValue)
+            {
+                config.UseDualstackEndpoint = defaultConfig.UseDualstackEndpoint.Value;
+            }
+            if (defaultConfig.UseFIPSEndpoint.HasValue)
+            {
+                config.UseFIPSEndpoint = defaultConfig.UseFIPSEndpoint.Value;
+            }
+            if (defaultConfig.UseHttp.HasValue)
+            {
+                config.UseHttp = defaultConfig.UseHttp.Value;
             }
 
-            // Setting RegionEndpoint only if ServiceURL was not set, because ServiceURL value will be lost otherwise
-            if (options.Region != null && string.IsNullOrEmpty(defaultConfig.ServiceURL))
+            if (defaultConfig.ServiceSpecificSettings.Count > 0)
             {
-                config.RegionEndpoint = options.Region;
+                ProcessServiceSpecificSettings(config, defaultConfig.ServiceSpecificSettings);
             }
 
             return config;
         }
+
+        private void ProcessServiceSpecificSettings(ClientConfig clientConfig, IDictionary<string, string> serviceSettings)
+        {
+            var singleArray = new object[1];
+            var properties = clientConfig.GetType().GetProperties(BindingFlags.Public | BindingFlags.Instance);
```

Comment:
I assume you could add something like 
```csharp
_ = typeof(AmazonS3Config).GetProperties();
```
to `CreateDefaultClientConfig` to force to preserve all `Config` properties, and then add UnconditionalSuppressMessage (knowing that properties are preserved).
Not sure if there's a significantly better way.

Reply:
I was surprised that the `GetProperties` method didn't have any attributes for requiring `DynamicallyAccessedMembers` or `RequiresUnreferencedCode` but I'll double check in action. My intention was to add `UnconditionalSuppressMessage` with defensive code and consider setting service specific settings a best effort in a trimmed environment. There is no way for this package to force the service specific configs not to be trimmed. Although since request pipeline in the service client is going to call of those settings at some point I don't see it likely that would be trimmed.

Reply:
@Dreamescaper you were right it did still trigger a warning. I have added the suppression as I mentioned before. I can confirm I can build and run a Native AOT with AWSSDK.Extensions.NETCore.Setup with no warnings.

Reply:
> Although since request pipeline in the service client is going to call of those settings at some point I don't see it likely they would be trimmed.

Per my understanding, even if the properties themselves are not trimmed, the reflection metadata still might, there they won't be returned in GetProperties call. 
E.g. this code prints "False", even though the property is crearly used:
```csharp
var config = new AmazonS3Config();
config.UseAccelerateEndpoint = true;
_ = config.UseAccelerateEndpoint;

var props = GetProperties(config);
Console.WriteLine(props.Contains("UseAccelerateEndpoint"));

string[] GetProperties(ClientConfig config)
{
    return config.GetType().GetProperties().Select(p => p.Name).ToArray();
}
```

> There is no way for this package to force the service specific configs not to be trimmed.
No way for _this_ package,  but it is possible from _other_ packages.
You can add the attribute similar to
`[DynamicDependency(DynamicallyAccessedMemberTypes.PublicProperties, typeof(AmazonS3Config))]`
to the `CreateDefaultClientConfig` methods you're generating.

Reply:
Awesome thanks! I didn't know about the `DynamicDependencyAttribute`. Latest commit adds that to generator  https://github.com/aws/aws-sdk-net/pull/3353/commits/7db32a3cffe8dfcba0de6d731966c9a78ddf4019